### PR TITLE
Allow adding extra dispatch keys to wrapper tensor subclass

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108316
* #108315
* #108314
* __->__ #108313

### CODE REVIEW

File: torch/autograd/grad_mode.py [hunk 1]
```diff
@@ -253,10 +254,10 @@ def __init__(self, mode: bool = True) -> None:
         self._inference_mode_raii_context: Optional[torch._C._InferenceMode] = None
         self.mode = mode
 
-    def __new__(cls, mode_or_func=True):
-        if isinstance(mode_or_func, bool):
+    def __new__(cls, mode=True):
+        if isinstance(mode, bool):
             return super().__new__(cls)
-        return cls()(mode_or_func)
+        return cls()(mode)
```
Comment (bdhirsh) [inline torch/autograd/grad_mode.py:260]: this is technically BC-breaking right?

**Comment by bdhirsh** [comment]
Also I think this change doesn't match the PR title?

**Comment by soulitzer** [comment]
> Also I think this change doesn't match the PR title?

Yeah no idea what happened to the stack lol, but force pushing ghstack fixes it. I've also updated it to be underscore.

File: torch/autograd/grad_mode.py (hunk 1 above)
Comment (soulitzer) [reply torch/autograd/grad_mode.py]: yup

**Comment by soulitzer** [comment]
ghstack really wanted to create another PR for some reason https://github.com/pytorch/pytorch/pull/108808
that one has the latest changes, so I'm closing this one

**Pull request closed without merging**